modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaService.java

```diff
@@ -93,33 +94,38 @@ private <R> CompletableFuture<R> sendToReplica(ClusterNode node, ReplicaRequest
             return (CompletableFuture<R>) replica.processRequest(req);
         }
 
-        return messagingService.invoke(node.address(), req, RPC_TIMEOUT).handle((response, throwable) -> {
+        CompletableFuture<R> res = new CompletableFuture<>();
+
+        messagingService.invoke(node.address(), req, RPC_TIMEOUT).handle((response, throwable) -> {
             if (throwable != null) {
                 if (throwable instanceof TimeoutException) {
-                    throw new ReplicationTimeoutException(req.groupId());
+                    res.completeExceptionally(new ReplicationTimeoutException(req.groupId()));
                 }
 
-                throw new ReplicationException(req.groupId(), throwable);
+                res.completeExceptionally(withCause(
+                        ReplicationException::new,
+                        REPLICA_COMMON_ERR,
+                        "Failed to process replica request [replicaGroupId=" + req.groupId() + ']',
+                        throwable));
             } else {
-                assert response instanceof ReplicaResponse : IgniteStringFormatter.format("Unexpected message response [resp={}]",
-                        response);
+                assert response instanceof ReplicaResponse : "Unexpected message response [resp=" + response + ']';
 
                 if (response instanceof TimestampAware) {
                     clock.update(((TimestampAware) response).timestamp());
                 }
 
                 if (response instanceof ErrorReplicaResponse) {
                     var errResp = (ErrorReplicaResponse) response;
-
-                    throw ExceptionUtils
-                            .error(errResp.errorTraceId(), errResp.errorCode(), errResp.errorClassName(), errResp.errorMessage(),
-                                    errResp.errorStackTrace());
+                    res.completeExceptionally(errResp.throwable());
                 } else {
-                    return (R) ((ReplicaResponse) response).result();
+                    res.complete((R) ((ReplicaResponse) response).result());
                 }
-
             }
+
+            return null;
```

Comment:
Well, I believe it's rather handle with return, or whenComplete with outer future, meaning that it's possible to remove return null in case of handle -> whenComplete.